### PR TITLE
Timetable and User business logic

### PR DESCRIPTION
**Pull request by Michal-MK**

### CODE REVIEW

File: business/src/test/java/cz/idomatojde/UserServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+import static org.mockito.ArgumentMatchers.anyInt;
+import static org.mockito.ArgumentMatchers.anyLong;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoInteractions;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
```
Comment (JirkaVrbka) [inline business/src/test/java/cz/idomatojde/UserServiceTest.java]: unused import

File: business/src/test/java/cz/idomatojde/facade/TimetableFacadeTest.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
+package cz.idomatojde.facade;
+
+import cz.idomatojde.dto.timetable.AddTimetableDTO;
+import cz.idomatojde.dto.timetable.TimetableEntryDTO;
+import cz.idomatojde.dto.user.RegisterUserDTO;
+import cz.idomatojde.dto.timetable.TimetableDTO;
```
Comment (JirkaVrbka) [inline business/src/test/java/cz/idomatojde/facade/TimetableFacadeTest.java:5]: unused import

File: business/src/test/java/cz/idomatojde/facade/TimetableFacadeTest.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.TestExecutionListeners;
+import org.springframework.test.context.testng.AbstractTestNGSpringContextTests;
+import org.springframework.test.context.transaction.TransactionalTestExecutionListener;
+import org.testng.annotations.Test;
+
+import javax.inject.Inject;
+import javax.transaction.Transactional;
+
+import java.util.Calendar;
```
Comment (JirkaVrbka) [inline business/src/test/java/cz/idomatojde/facade/TimetableFacadeTest.java]: unused import

File: dao/src/main/java/cz/idomatojde/dao/OfferDao.java [hunk 4]
```diff
@@ -20,6 +20,14 @@
      */
     List<Offer> findByUser(User user);
 
+    /**
+     * Finds all offers a {@link User} is sibscribed to
```
Comment (JirkaVrbka) [inline dao/src/main/java/cz/idomatojde/dao/OfferDao.java]: Typo in "sibscribed"

File: api/src/main/java/cz/idomatojde/facade/TimetableFacade.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * @author Michal Hazdra
+ */
+public interface TimetableFacade {
+
+    /**
+     * Create a new timetable
+     *
+     * @param timetableDTO necessary timetable creation information
+     * @return uniqe id of the timetable
```
Comment (JirkaVrbka) [inline api/src/main/java/cz/idomatojde/facade/TimetableFacade.java]: typo in "uniqe"

File: api/src/main/java/cz/idomatojde/facade/UserFacade.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+ */
+public interface UserFacade {
+
+    /**
+     * Register a new user
+     *
+     * @param registrationInfo necessary user creation information
+     * @return the ID of the User
+     */
+    Long registerUser(RegisterUserDTO registrationInfo);
```
Comment (JirkaVrbka) [inline api/src/main/java/cz/idomatojde/facade/UserFacade.java:21]: It is not necessary to return Long, add Id to the DTO (as reference project does)

File: business/src/main/java/cz/idomatojde/services/TimetableServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+
+    @Override
+    public void removeEntry(TimetableEntry entry) {
+        timetables.removeEntry(entry);
+    }
+
+
+    @Override
+    public void updateEntry(TimetableEntry entry) {
```
Comment (JirkaVrbka) [inline business/src/main/java/cz/idomatojde/services/TimetableServiceImpl.java]: Updating is done automatically, no need for this method

File: business/src/main/java/cz/idomatojde/services/facades/TimetableFacadeImpl.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+    @Inject
+    public TimetableFacadeImpl(TimetableService timetableService, UserService userService, MappingService mapService) {
+        this.timetableService = timetableService;
+        this.userService = userService;
+        this.mapService = mapService;
+    }
+
+    @Override
+    public long addTimetable(AddTimetableDTO timetableDTO) {
```
Comment (JirkaVrbka) [inline business/src/main/java/cz/idomatojde/services/facades/TimetableFacadeImpl.java:35]: Same as comment before, add Id to DTO

File: business/src/main/java/cz/idomatojde/services/facades/UserFacadeImpl.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        User u = userService.getById(userId);
+
+        return new UserCreditsDTO(u.getCredits(), u.getBonusCredits());
+    }
+
+    @Override
+    public void changePhoneNumber(long userId, String phoneNumber) {
+        User u = userService.getById(userId);
+
+        // TODO Phone number verification
```
Comment (JirkaVrbka) [inline business/src/main/java/cz/idomatojde/services/facades/UserFacadeImpl.java]: TODO?

File: business/src/test/java/cz/idomatojde/TimetableServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,364 @@
...
+public class TimetableServiceTest extends AbstractTestNGSpringContextTests {
+
+    private TimetableDAO mockDao;
+
+    private TimetableService service;
+
+    private final LocalTime TIME_4PM = LocalTime.of(16, 0);
+    private final Duration DUR_2H = Duration.ofHours(2);
+
+    private User DEF_USER;
```
Comment (JirkaVrbka) [inline business/src/test/java/cz/idomatojde/TimetableServiceTest.java]: Why upper cases?

File: business/src/test/java/cz/idomatojde/UserServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+ */
+@ContextConfiguration("classpath:applicationConfig.xml")
+@TestExecutionListeners(TransactionalTestExecutionListener.class)
+public class UserServiceTest extends AbstractTestNGSpringContextTests {
+
+    private UserDao mockDao;
+
+    private UserService service;
+
+    private User DEF_USER;
```
Comment (JirkaVrbka) [inline business/src/test/java/cz/idomatojde/UserServiceTest.java]: Why upper cases?

File: api/src/main/java/cz/idomatojde/facade/UserFacade.java (hunk 6 above)
Comment (Michal-MK) [reply api/src/main/java/cz/idomatojde/facade/UserFacade.java:21]: IMO, the DTO should contain only the info necessary for registration, the User's ID is not necessary for a registration to succeed.

Comment (JirkaVrbka) [reply api/src/main/java/cz/idomatojde/facade/UserFacade.java:21]: I would refer you to reference project. They have it like that

Comment (Michal-MK) [reply api/src/main/java/cz/idomatojde/facade/UserFacade.java:21]: Are you sure? [Link](https://i.gyazo.com/3d240ae82045388a401c9c62df37a2cf.png)

Comment (JirkaVrbka) [reply api/src/main/java/cz/idomatojde/facade/UserFacade.java:21]: [https://github.com/fi-muni/PA165/blob/master/eshop-service/src/main/java/cz/fi/muni/pa165/service/facade/UserFacadeImpl.java](link)

They have inconsistency as well. Let me change the code and return Id. Actually, I like it more

**Review by JirkaVrbka — APPROVED** [review]

File: api/src/main/java/cz/idomatojde/facade/UserFacade.java (hunk 6 above)
Comment (Michal-MK) [reply api/src/main/java/cz/idomatojde/facade/UserFacade.java:21]: Interesting, I just continued with what we did in DAO and Service layers to be consistent.

**Pull request merged**